x-pack/legacy/plugins/maps/public/layers/styles/vector/components/color/dynamic_color_form.js

```diff
@@ -7,56 +7,174 @@
+import { ColorMapSelect } from './color_map_select';
 import { EuiFlexGroup, EuiFlexItem, EuiSpacer } from '@elastic/eui';
+import { CATEGORICAL_DATA_TYPES, COLOR_MAP_TYPE } from '../../../../../../common/constants';
+import { OrdinalFieldMetaOptionsPopover } from '../ordinal_field_meta_options_popover';
+import { FormattedMessage } from '@kbn/i18n/react';
+import { COLOR_GRADIENTS, COLOR_PALETTES_INPUTS } from '../../../color_utils';
 
-export function DynamicColorForm({
-  fields,
-  onDynamicStyleChange,
-  staticDynamicSelect,
-  styleProperty,
-}) {
-  const styleOptions = styleProperty.getOptions();
-
-  const onFieldChange = ({ field }) => {
-    onDynamicStyleChange(styleProperty.getStyleName(), { ...styleOptions, field });
+export class DynamicColorForm extends React.Component {
+  state = {
+    colorMapType: COLOR_MAP_TYPE.ORDINAL,
   };
 
-  const onColorChange = colorOptions => {
-    onDynamicStyleChange(styleProperty.getStyleName(), {
-      ...styleOptions,
-      ...colorOptions,
-    });
-  };
+  constructor() {
+    super();
+    this._isMounted = false;
+  }
 
-  let colorRampSelect;
-  if (styleOptions.field && styleOptions.field.name) {
-    colorRampSelect = (
-      <ColorRampSelect
-        onChange={onColorChange}
-        color={styleOptions.color}
-        customColorRamp={styleOptions.customColorRamp}
-        useCustomColorRamp={_.get(styleOptions, 'useCustomColorRamp', false)}
-        compressed
-      />
-    );
+  componentWillUnmount() {
+    this._isMounted = false;
+  }
+
+  componentDidMount() {
+    this._isMounted = true;
+    this._loadColorMapType();
+  }
+
+  componentDidUpdate() {
+    this._loadColorMapType();
+  }
+
+  async _loadColorMapType() {
+    const field = this.props.styleProperty.getField();
+    if (!field) {
+      return;
+    }
+    const dataType = await field.getDataType();
+    const colorMapType = CATEGORICAL_DATA_TYPES.includes(dataType)
+      ? COLOR_MAP_TYPE.CATEGORICAL
+      : COLOR_MAP_TYPE.ORDINAL;
+    if (this._isMounted && this.state.colorMapType !== colorMapType) {
+      this.setState({ colorMapType });
+    }
   }
 
-  return (
-    <Fragment>
-      <EuiFlexGroup gutterSize="none" justifyContent="flexEnd">
-        <EuiFlexItem grow={false}>{staticDynamicSelect}</EuiFlexItem>
-        <EuiFlexItem>
-          <FieldSelect
-            fields={fields}
-            selectedFieldName={_.get(styleOptions, 'field.name')}
-            onChange={onFieldChange}
+  _onFieldMetaOptionsChange = fieldMetaOptions => {
+    const options = {
+      ...this.props.styleProperty.getOptions(),
+      fieldMetaOptions,
+    };
+    this.props.onDynamicStyleChange(this.props.styleProperty.getStyleName(), options);
+  };
+
+  _getColorSelector() {
+    const { onDynamicStyleChange, styleProperty } = this.props;
+    const styleOptions = styleProperty.getOptions();
+
+    let colorSelect;
+    if (styleOptions.field && styleOptions.field.name) {
```

Comment:
Can you just return early if not field and field name to avoid this extra nesting.